### DISCUSSION TITLE
Help debug strange behavior

### DISCUSSION THREAD

**Discussion by cbblanke** [NASA-LIS/LISF#889] [Q&A] [answered]
I am using LIS 7.1 to do Noah forecast runs with GFS forcing in Africa.  They are initialized with a near-real-time run using GDAS (from a run on a different machine).  We have 2 existing domains and it works flawlessly.  I tried to create a bigger domain (all of Africa) and the first part (GDAS forcing) runs fine.  Then the second part (GFS forcing) crashes almost immediately.  This behavior is really difficult to debug.  Here are some things I have found.

- The last thing it prints (on every processor) is "reading mask  588 319".  This happens in LIS_fileIOMod.F90
- The actual line where it dies (Fortran debugging) is where I try to print the name of LIS_rc%paramfile(n)
- If I remove that line that I added, it dies where it tries to inquire about the file.
-  The paramfile exists and is readable, and the name is not too long for the variable.

Here are the really strange behaviors.
- Many things I try to write to LIS_logunit for diagnostic purposes only go to fort.13.  (even though the lisdiag.???? files are getting other output).  That includes a message when I go into readConfig and a message in readConfig that identifies the paramfile.  
- I am sure I have successfully recompiled LIS and am using that executable.

Any idea what is going on?  The nowcast and the forecast runs both work for the two original domains.  For the new domain, the nowcast runs but not the forecast.

**Comment by emkemp** [comment]
Hi @cbblanke 

I'm speculating, but it may be running out of virtual memory.  Do you see a segmentation fault error in your output?  (If you're running in a batch queue environment, it would be in a separate file produced by the queuing manager -- PBS, SLURM, etc.)

**Reply by cbblanke** [reply]
Yes, it says "forrtl: severe (174): SIGSEGV, segmentation fault occurred"  (I've never known what that meant so I guess I ignored it.)

How do I find out how much virtual memory I need?  Do I just have to do trial and error?  This is on a virtual machine, it's possible the memory could be increased but I'll have to check (and I don't know how much to ask for).  Alternatively we can reduce the domain size some as I wanted to get all of Africa but it's not a requirement.

**Reply by emkemp** [reply]
(I apologize @cbblanke, I'm afraid this fell off my radar.)

I don't have a good suggestion for assessing memory requirements other than trial and error.  However, I do suggest setting 'ulimit -s unlimited' in your environment to maximize the stack space available for automatic arrays.

**Reply by cbblanke** [reply, marked as answer]
It does look like "ulimit -s unlimited" did the trick.  I was running the second part on a different machine, which I should have mentioned.  But not something I knew to check.  Thanks!

**Reply by emkemp** [reply]
Excellent!

A postmortem:  The reason 'ulimit -s unlimited' is important is that parts of LIS use automatic arrays, which are allocated with the size passed as a subroutine or function argument.  Example code:

subroutine demo(nobs)
  implicit none
  integer, intent(in) :: nobs
  
  real :: array1(nobs)
  real :: array2(nobs)
  
  ....
  
end subroutine demo

By default, Fortran programs will automatically allocate on the computer "stack", a segment of memory that is reserved for such uses. (Deallocation is handled automatically when the subroutine ends.)  The size of the stack varies by system and can be adjusted by that ulimit setting.  If it's too small, the array allocation won't work, and the segmentation fault occurs.  (Declaring the arrays as "allocatable" and using explicit "allocate" and "deallocate" statements use the memory on the system "heap", which is usually much larger.  Perhaps in the longer term we should switch to all heap allocations in LIS, but there's a lot of code to change to make that happen.)

So now you know... the rest of the story.   Always set 'ulimit -s unlimited' (or whatever the equivalent is if you're using a different shell than bash.)

**Reply by cbblanke** [reply]
So is there a drawback to that setting?  Why is it not the default?

**Reply by emkemp** [reply]
I'm not sure about drawbacks.  As for why it's small by default, I suspect it's a legacy of C (which was used to implement UNIX and Linux).

The "stack" and "heap" are just different parts of the total machine memory.  The stack is used in a "last in, first out" order to keep track of which subroutines/functions are active as well as their local variables.  When a subroutine returns, memory is freed.  In contrast, the heap is meant to be carved up as needed.

In C, there is no concept of an automatic array with dimensions set at runtime.  If you need to set aside memory, you must use an explicit allocator (malloc), and save the starting address of that memory in a pointer.  And if you want to make that memory available to another function, you must again pass a pointer.  So in practice, the pointers (and all the other local variables) usually takes up a relatively small amount of stack space.  Since UNIX was C-based, it made sense to set stack space limits small, and set heap limits large. And the defaults continues.

But Fortran is not C.  In fact, prior to Fortran 90 there was no concept of an allocatable array, and they either used hardwired dimensions or were automatic arrays.  The use of stack or heap in Fortran is left to the implementor, and most (all?) use the the stack for automatic arrays.  (The Intel compiler has a flag to force use of the heap for automatic arrays, but I don't believe gfortran does.)

Anyway, that's my theory.  But I'm not a computer scientist or OS guru.

**Reply by cbblanke** [reply]
OK, thanks.
